### PR TITLE
[ISSUE #4857] Fix broker can't auto register to controller again

### LINKED ISSUE #4857: Broker can't auto register to controller again in some case
ReplicasManager#registerBrokerToController method is used in three places
image
image
as following picture:
image
when throw exception ReplicasManager#registerBrokerToController  will not to execute.
in controller mode, ReplicasInfoManager#getReplicaInfo as following 
image
if `isContainsBroker(brokerName) ` result return false then return result code is ResponseCode.CONTROLLER_INVALID_REQUEST,
In this case, broker should register to controller again

### PR DESCRIPTION
**Pull request by mxsm**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

Fix #4857 

## Brief changelog

- Fix broker can't auto register to controller again

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java [hunk 1]
```diff
@@ -347,6 +351,12 @@ private void schedulingSyncBrokerMetadata() {
                 }
             } catch (final Exception e) {
                 LOGGER.warn("Error happen when get broker {}'s metadata", this.brokerConfig.getBrokerName(), e);
+                if (e instanceof MQBrokerException) {
+                    int code = ((MQBrokerException) e).getResponseCode();
+                    if (code == CONTROLLER_INVALID_REQUEST || code == CONTROLLER_NOT_LEADER) {
```
[the remark this answers is not in the document]
Comment (hzh0425) [reply broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java:356]: I think you are right

[the remark this answers is not in the document]
Comment (mxsm) [reply broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java:356]: I will adding a new response code to indicate this situation.  and ReplicasInfoManager method to process different request code, if nothing to do will return ResponseCode.CONTROLLER_INVALID_REQUEST code. I think different handle should return coding related to processing. I will submit a PR to deal with in the future.

[the remark this answers is not in the document]
Comment (mxsm) [reply broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java:356]: > if code == CONTROLLER_NOT_LEADER, we don't need to do anything.

OK, I got it

File: controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java [hunk 2]
```diff
@@ -288,7 +288,7 @@ public ControllerResult<GetReplicaInfoResponseHeader> getReplicaInfo(final GetRe
             result.setBody(new SyncStateSet(syncStateInfo.getSyncStateSet(), syncStateInfo.getSyncStateSetEpoch()).encode());
             return result;
         }
-        result.setCodeAndRemark(ResponseCode.CONTROLLER_INVALID_REQUEST, "Broker metadata is not existed");
+        result.setCodeAndRemark(ResponseCode.CONTROLLER_INVALID_GETREPLICAINFO, "Broker metadata is not existed");
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java:291]: How about CONTROLLER_BROKER_MEATA_NOT_EXIST？

Comment (mxsm) [reply controller/src/main/java/org/apache/rocketmq/controller/impl/manager/ReplicasInfoManager.java:291]: Good idea

File: broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java [hunk 3]
```diff
@@ -347,6 +351,12 @@ private void schedulingSyncBrokerMetadata() {
                 }
             } catch (final Exception e) {
                 LOGGER.warn("Error happen when get broker {}'s metadata", this.brokerConfig.getBrokerName(), e);
+                if (e instanceof MQBrokerException) {
+                    int code = ((MQBrokerException) e).getResponseCode();
+                    if (code == CONTROLLER_INVALID_GETREPLICAINFO) {
+                        registerBrokerToController();
```
Comment (ShannonDing) [inline broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java:357]: how about try to register again after a while?

Comment (mxsm) [reply broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java:357]: > how about try to register again after a while?

That's good idea

File: broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java [hunk 4]
```diff
@@ -347,6 +351,17 @@ private void schedulingSyncBrokerMetadata() {
...
+                    int code = ((MQBrokerException) e).getResponseCode();
+                    if (code == CONTROLLER_BROKER_METADATA_NOT_EXIST) {
+                        try {
+                            registerBrokerToController();
+                            TimeUnit.SECONDS.sleep(2);
+                        } catch (InterruptedException ignore) {
+
+                        }
+                    }
+                }
```
Comment (RongtongJin) [inline broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java:364]: How about catching  MQBrokerException before Exception？ In addition, could you rebase or merge the latest develop branch to this pull request？ We add new rules for merging

Comment (mxsm) [reply broker/src/main/java/org/apache/rocketmq/broker/controller/ReplicasManager.java:364]: yes I will do this

**Review by hzh0425 — APPROVED** [review]
LGTM

**Review by RongtongJin — APPROVED** [review]

**Pull request merged**